exec/cnex/array.c

```diff
@@ -135,14 +152,17 @@ size_t array_appendElement(Array *self, const Cell *element)
         fatal_error("Uninitialized array passed to array_appendElement!");
     }
     if (self->data) {
-        self->data = realloc(self->data, sizeof(Cell) * (self->size + 1));
-        if (self->data == NULL) {
-            fatal_error("Not enough memory to expand array.");
+        if (self->max <= self->size) {
+            self->max *= 2;
+            self->data = realloc(self->data, sizeof(Cell) * self->max);
+            if (self->data == NULL) {
+                fatal_error("Not enough memory to expand array.");
+            }
         }
         self->size++;
     }
     if (self->data == NULL) {
-        self->data = malloc(sizeof(Cell));
+        self->data = malloc(sizeof(Cell) * self->max);
```

Comment:
Under what conditions is `self->data` NULL, but `self->max` is nonzero? What does a nonzero max mean if there is no data? What if `self->max` is zero here? Line 169 seems to assume it's at least 1.

Reply:
So, this was a unique scenario.  This happened in the startup routine of cnex, building the arguments, for one.  A call to `cell_createCellArray(0)` is made, then each arg is added to the array with a call to `cell_arrayAppendElement()`.  This was a case of `self->data` being NULL, as was `self->max`.  All of this code has been reworked based on the review comments so the above description will no longer happen.  Now, `self->data` should _never_ be NULL.